### PR TITLE
Add owl.0.2.9

### PR DESCRIPTION
**Pull request by ryanrhymes**
- add new neurons in `Neural` module
- add checkpoint function in `Neural` module
- new `Optimise` module built completely on `Algodiff`
- Many core functions are enhanced, e.g. `broadcast` functions.

This submission is mostly for the purpose of archiving the last release of Owl (based on Oasis and with many old designs). In future releases, many breaking changes will be introduced including its APIs, internal designs, build system, and etc.

### CODE REVIEW

File: packages/owl/owl.0.2.9/opam [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+]
+build-test: [
+  ["oasis" "setup"]
+  ["ocaml" "setup.ml" "-configure" "--enable-tests"]
+  ["ocaml" "setup.ml" "-build"]
+  ["ocaml" "setup.ml" "-test"]
+]
+build-doc: [ "ocaml" "setup.ml" "-doc" ]
+depends: [
+  "ctypes"
```
Comment (avsm) [inline packages/owl/owl.0.2.9/opam]: There is a missing build dependency on atdgen

```
#=== ERROR while installing owl.0.2.9 =========================================#
# opam-version         1.2.2 (58ef3b8213100953848d362f7120a30356d7f77d)
# os                   linux
# command              ocaml setup.ml -configure --prefix /home/opam/.opam/4.04.2 --bindir /home/opam/.opam/4.04.2/bin
# path                 /home/opam/.opam/4.04.2/build/owl.0.2.9
# compiler             4.04.2
# exit-code            1
# env-file             /home/opam/.opam/4.04.2/build/owl.0.2.9/owl-29-eeb389.env
# stdout-file          /home/opam/.opam/4.04.2/build/owl.0.2.9/owl-29-eeb389.out
# stderr-file          /home/opam/.opam/4.04.2/build/owl.0.2.9/owl-29-eeb389.err
### stderr ###
# ocamlfind: Package `atdgen' not found
# W: Field 'pkg_atdgen' is not set: Command ''/home/opam/.opam/4.04.2/bin/ocamlfind' query -format %d atdgen > '/tmp/oasis-181b8b.txt'' terminated with error code 2
# E: Cannot find findlib package atdgen
# E: Failure("1 configuration error")
```

**Comment by ryanrhymes** [comment]
It only built successfully on macos, the error is as follows.

```bash
Error: Error on dynamically loaded library: ./lib/dllowl_stubs.so: ./lib/dllowl_stubs.so: undefined symbol: LAPACKE_sgghd3
....
```

There seems to be a lot of challenges in linking to cblas/lapacke on linux. I think the cblas/lapacke version is a bit outdated and missing some new functions used in Owl.

**Comment by yallop** [comment]
I've  manually merged this into `master` (541a7791).

**Comment by ryanrhymes** [comment]
Many thanks @yallop

**Pull request closed without merging**